src/Features/LanguageServer/Protocol/LspServices/LspServices.cs

```diff
@@ -59,90 +81,132 @@ internal sealed class LspServices : ILspServices
+        {
+            return this;
+        }
 
         // Check the base services first
-        if (_baseServices.TryGetValue(type, out var baseServices))
+        if (_baseServices.TryGetValue(typeName, out var baseServices))
         {
-            lspService = baseServices.Select(creatorFunc => creatorFunc(this)).SingleOrDefault();
-            if (lspService is not null)
-            {
-                return lspService;
-            }
+            // It's possible that there may be more than one base service registered for the same type,
+            // such as IMethodHandler. If that's the case, we return null.
+            return baseServices is [var baseService]
+                ? baseService.GetInstance(this)
+                : null;
         }
 
-        if (_lazyMefLspServices.TryGetValue(type, out var lazyService))
+        if (_lazyMefLspServices.TryGetValue(typeName, out var lazyService))
         {
             // If we are creating a stateful LSP service for the first time, we need to check
             // if it is disposable after creation and keep it around to dispose of on shutdown.
             // Stateless LSP services will be disposed of on MEF container disposal.
             var checkDisposal = !lazyService.Metadata.IsStateless && !lazyService.IsValueCreated;
 
-            lspService = lazyService.Value;
+            var lspService = lazyService.Value;
             if (checkDisposal && lspService is IDisposable disposable)
             {
                 lock (_gate)
                 {
-                    var res = _servicesToDispose.Add(disposable);
+                    _servicesToDispose.Add(disposable);
                 }
             }
 
             return lspService;
         }
 
-        lspService = null;
-        return lspService;
+        return null;
     }
 
-    public ImmutableArray<Type> GetRegisteredServices()
-        => _lazyMefLspServices.Keys.ToImmutableArray();
-
-    public bool SupportsGetRegisteredServices()
+    public ImmutableArray<(IMethodHandler? Instance, TypeRef HandlerTypeRef, ImmutableArray<HandlerMethodDetails> Methods)> GetMethodHandlers()
     {
-        return true;
-    }
+        using var _ = ArrayBuilder<(IMethodHandler?, TypeRef, ImmutableArray<HandlerMethodDetails>)>.GetInstance(out var builder);
 
-    private IEnumerable<T> GetBaseServices<T>()
-        => _baseServices.TryGetValue(typeof(T), out var baseServices)
-            ? baseServices.Select(creatorFunc => (T)creatorFunc(this)).ToImmutableArray()
-            : (IEnumerable<T>)[];
+        // First, add any IMethodHandlers found in base services.
+        foreach (var handler in GetBaseServices<IMethodHandler>())
+        {
+            var handlerType = handler.GetType();
+            var methods = HandlerMethodDetails.From(handlerType);
 
-    private IEnumerable<T> GetMefServices<T>()
-    {
-        if (typeof(T) == typeof(IMethodHandler))
+            builder.Add((handler, handlerType, methods));
```

Comment:
do we need to specially handle and return the instance here?  In the caller it looks like if there is no instance it calls back into `lspServices.TryGetService(handlerTypeRef.GetResolvedType(), out var lspService)` - does that not work for the base service here?

Reply:
I'm not sure I followed your question. This is the implementation of `IMethodHandlerProvider.GetMethodHandlers()`, which returns all `IMethodHandler` implementations, whether they are already created as base services or lazy MEF services. Things like `InitializeHandler` are already instantiated and there's no reason to create them again.

Reply:
In the caller here - https://github.com/dotnet/roslyn/pull/73566/files/beca74459d4f3a1ea08afb78479a8164fcce9b99#diff-5a180c92657a4c9f9f5e9779e5189cc840cf93baf9c93a2dfbb7f6b9e7c6483dR69

it doesn't seem like there is  a huge need to differentiate whether or not it was already created - seems like we could always call `GetLazyHandlerFromTypeRef` since that just calls back to get the LSP service.  Then the caller of `GetMethodHandlers` doesn't need to care if the instance is created or not

Reply:
Oh, I see what you mean now. Yeah, I can make that change.